Enable dev services for all named mongodb-client connections

evanchooly opened the pull request:
fixes #25278

Commit:
Enable dev services for all named mongodb-client connections

fixes #25278

geoand commented:
> Could we infer the existence of a named client by discovering uses of @MongoClientName and backfilling that client using only that name and the default values of the config? In essence, lazy init'ing the named client and letting devservices have its usual go at it

Probably, but I am thinking that it might break stuff, so I would like to avoid it.

geoand commented on integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java:
```diff
@@ -9,17 +9,20 @@
 public class BookResource {
 
     @Inject
+    @MongoClientName("parameter-injection")
```
I think we need the test to have both named and non-named clients

evanchooly commented:
> > Could we infer the existence of a named client by discovering uses of @MongoClientName and backfilling that client using only that name and the default values of the config? In essence, lazy init'ing the named client and letting devservices have its usual go at it
> 
> Probably, but I am thinking that it might break stuff, so I would like to avoid it.

actually, the more i think about this the less i like it.  typos would be a nightmare.  (even though those names should be held in constants somewhere. ;) )

evanchooly commented:
*bump*

geoand commented:
The affected test is **not** using dev services as the connection string has been set - mongo is being spun up by `io.quarkus.test.mongodb.MongoTestResource`

evanchooly commented on integration-tests/mongodb-client/src/main/resources/application.properties:
```diff
@@ -1,4 +1,7 @@
+quarkus.mongodb.parameter-injection.write-concern.journal=true
+
+quarkus.mongodb.dev-services.write-concern.journal=false
+quarkus.mongodb.dev-services2.write-concern.journal=false
```
what do you think of this @geoand?  And the `BookResource` change above.  It's not the "cleanest" but I was trying to find a simple way to test both the `MongoTestResource` driven stuff and the dev services as well and this is the simplest I could make it.

geoand replied:
I like the idea.

geoand commented on integration-tests/mongodb-client/src/main/java/io/quarkus/it/mongodb/BookResource.java:
```diff
@@ -9,36 +9,60 @@
     @POST
     public Response addBook(Book book) {
         getCollection().insertOne(book);
+        getServicedCollection().insertOne(book);
```
We aren't querying this collection anywhere, are we? I think we should close the loop (somewhere during the duration of the test) and also do that

evanchooly replied:
smart.  just pushed that change.

geoand reviewed (approved):
LGTM!

evanchooly commented:
> Hm... The test failure is suspicious

No docker on CI ...

loicmathieu commented:
@evanchooly there is no docker on Windows CI. That's why we don't use testcontainers for MongoDB tests but the testresource. But even the testresource didn't works very well on Windows due to some issue on flapdoodle side so we ends up disabling the test on Windows: `@DisabledOnOs(WINDOWS)`.

loicmathieu commented:
> @evanchooly there is no docker on Windows CI. That's why we don't use testcontainers for MongoDB tests but the testresource. But even the testresource didn't works very well on Windows due to some issue on flapdoodle side so we ends up disabling the test on Windows: @DisabledOnOs(WINDOWS).

But you didn't add new tests so I don't understand why you have this ...

geoand commented:
> But you didn't add new tests so I don't understand why you have this ...

Probably because:

```xml
            <plugin>
                <artifactId>maven-failsafe-plugin</artifactId>
                <configuration>
                    <skip>true</skip>
                </configuration>
            </plugin>
```

was removed

evanchooly commented:
fixed the windows failure by disabling it via a profile when on windows.  the last CI run was all green so I squashed the commits and force pushed.  we should be all good to go now.

famod commented:
It seems this PR has broken `NamedMongoClientConfigTest` and `NamedReactiveMongoClientConfigTest`.

The failsafe-skip-profile won't work for those as both are run via surefire.

geoand commented:
Yup, @evanchooly is looking into it
